folding-schemes/src/frontend/noname/mod.rs

```diff
@@ -0,0 +1,228 @@
+use ark_r1cs_std::alloc::AllocVar;
+use ark_r1cs_std::fields::fp::FpVar;
+use ark_relations::r1cs::{ConstraintSynthesizer, SynthesisError};
+use num_bigint::BigUint;
+use std::marker::PhantomData;
+
+use super::FCircuit;
+use ark_ff::PrimeField;
+use ark_noname::utils::compile_source_code;
+use noname::backends::{r1cs::R1CS as R1CSNoName, BackendField};
+use noname::witness::CompiledCircuit;
+
+pub mod utils;
+
+use utils::*;
+
+#[derive(Debug, Clone)]
+pub struct NoNameFCircuit<F: PrimeField, BF: BackendField> {
+    pub state_len: usize,
+    pub external_inputs_len: usize,
+    pub circuit: CompiledCircuit<R1CSNoName<BF>>,
+    _f: PhantomData<F>,
+}
+
+impl<F: PrimeField, BF: BackendField> FCircuit<F> for NoNameFCircuit<F, BF> {
+    type Params = (String, usize, usize);
+
+    fn new(params: Self::Params) -> Result<Self, crate::Error> {
+        let (code, state_len, external_inputs_len) = params;
+        let compiled_circuit = compile_source_code::<BF>(&code).map_err(|_| {
+            Error::Other("Encountered an error while compiling a noname circuit".to_owned())
+        })?;
+        Ok(NoNameFCircuit {
+            state_len,
+            external_inputs_len,
+            circuit: compiled_circuit,
+            _f: PhantomData,
+        })
+    }
+
+    fn state_len(&self) -> usize {
+        self.state_len
+    }
+
+    fn external_inputs_len(&self) -> usize {
+        self.external_inputs_len
+    }
+
+    fn step_native(
+        &self,
+        _i: usize,
+        z_i: Vec<F>,
+        external_inputs: Vec<F>,
+    ) -> Result<Vec<F>, crate::Error> {
+        // compute witness using noname witness generator and outputs the next step
+        let external_inputs_json = match !external_inputs.is_empty() {
+            false => None,
+            true => {
+                let json =
+                    serde_json::to_string(&NoNameJSONExternalInputs::from(external_inputs.clone()))
+                        .map_err(|e| Error::JSONSerdeError(e.to_string()))?;
+                Some(json)
+            }
+        };
+
+        let ivc_inputs = serde_json::to_string(&NoNameJSONIVCInput::from(z_i.clone()))
+            .map_err(|e| Error::JSONSerdeError(e.to_string()))?;
+        let noname_inputs = NoNameIVCInputs::<F>::try_from((ivc_inputs, external_inputs_json))
+            .map_err(|_| SynthesisError::Unsatisfiable)?;
+        let (ivc_inputs, noname_external_inputs) = noname_inputs.try_into()?;
+        let noname_witness = self
+            .circuit
+            .generate_witness(ivc_inputs, noname_external_inputs)
+            .map_err(|e| Error::WitnessCalculationError(e.to_string()))?;
+
+        let mut assigned_z_i1 = vec![];
+        let z_i1_end_index = z_i.len() + 1;
+        for idx in 1..z_i1_end_index {
+            let value: BigUint = Into::into(noname_witness.witness[idx]);
+            let field_element = F::from(value);
+            assigned_z_i1.push(field_element);
+        }
```

Comment:
Is Noname restricted to JSON?? If so, that's ok. Otherwise, I think we should use plain serde and allow the user to specify what does it want to use to deserialize stuff.

Reply:
You are right! I did a major refactor for this PR based on your remark. I removed most of the transformations I was doing. Now, the `generate_step_constraints` method is much more [lightweight](https://github.com/privacy-scaling-explorations/sonobe/blob/d2db27d434dd70c5078cbd1738a9e8e6e0577c4d/folding-schemes/src/frontend/noname/mod.rs#L75) and the content of the `utils.rs` file is [minimal](https://github.com/privacy-scaling-explorations/sonobe/blob/feat/noname-frontend/folding-schemes/src/frontend/noname/utils.rs).